anesthetic/plot.py

```diff
@@ -481,7 +481,12 @@ def kde_plot_1d(ax, data, *args, **kwargs):
         Sample weights.
 
     ncompress: int, optional
-        Degree of compression. Default 1000
+        Degree of compression.
+        Default 10000
```

Comment:
Should the default be `None`, which compresses 'optimally' to the channel capacity? 10000 seems a little arbitrary, and at this value makes one wonder if we even need compression at all! How does the speed compare with no compression on a cosmological run?.

Reply:
> Should the default be None, which compresses 'optimally' to the channel capacity? 10000 seems a little arbitrary, and at this value makes one wonder if we even need compression at all! 

Agreed on all accounts. We have a few decisions to make:
1) Should `nplot` be the same for 1D and 2D? I think it would be better to have separate `nplot_1d` and `nplot_2d`, since aesthetically they have very different minimum requirements and speedwise 1D is much faster (regardless of `ncompress`) than 2D.
2) Which default `ncompress` for 1D? From the timings I'd say to either go with channel capacity or with no compression at all.
3) Which default `ncompress` for 2D? No compression is definitely not an option, too slow. We could go with:
   * `ncompress=1000`: current default, good for quick look, for paper plot typically insufficient for non-standard-gaussian contours
   * `ncompress=int(ns.neff())`: potentially enough for paper plots, but maybe too slow (especially when plotting multiple contours)

Thoughts?



> How does the speed compare with no compression on a cosmological run?.

See timings below.



&nbsp;

### `kde_plot_1d`

For the 1D case `ncompress` is negligible for computation time for an 86042 sample cosmology run:
```python
ns.plot_2d(axes, kind='kde_1d', ...)
```
* `nplot=200, ncompress=1000`: 421ms +- 9ms
* `nplot=200, ncompress=10000`: 480ms +- 10ms
* `nplot=200, ncompress=int(ns.neff())`: 917ms +- 56ms   (`ncompress=16874`)
* `nplot=200, ncompress=100000`: 1290ms += 20ms   (`ncompress=86042`)



&nbsp;

### `kde_plot_2d`

For the 2D case `ncompress` is more important regarding computation time. For an 86042 sample cosmology run:
```python
ns.plot_2d(axes, kind='kde_2d', ...)
```
* `nplot=1000, ncompress=1000`: 3.09s +- 0.08s
* `nplot=1000, ncompress=10000`: 6.58s +- 0.20s
* `nplot=1000, ncompress=int(ns.neff())`: 9.29s +- 0.22s   (`ncompress=16874`)
* `nplot=1000, ncompress=100000`: 32.5s +- 0.30s   (`ncompress=86042`)

Reply:
Many thanks for these excellent data!

> 1\. Should `nplot` be the same for 1D and 2D? I think it would be better to have separate `nplot_1d` and `nplot_2d`, since aesthetically they have very different minimum requirements and speedwise 1D is much faster (regardless of `ncompress`) than 2D.

No. I agree.


> 2\. Which default `ncompress` for 1D? From the timings I'd say to either go with channel capacity or with no compression at all.

Dealer's choice here. Tempting to go with the less interfering of the two options (i.e. no compression).


> 3\. Which default `ncompress` for 2D? No compression is definitely not an option, too slow. We could go with:
>     
>     * `ncompress=1000`: current default, good for quick look, for paper plot typically insufficient for non-standard-gaussian contours
>     * `ncompress=int(ns.neff())`: potentially enough for paper plots, but maybe too slow (especially when plotting multiple contours)

Here my answer is a bit more involved.

- The current commit switches from a tricontour-based samples approach to a meshgrid-based one. 
- The advantage of the previous setup was that using samples to plot 1 and 2-sigma contours naturally focuses points on the region that we need (which meshgrid doesn't, particularly for correlated distributions). In principle, one would only need ~O(20) plotting samples to adequately populate 2-sigma contours. If we had this set-up we could set nplot for an 2 sigma plot to be 370 = (1/1-sigmas['3sigma']), or something proportional (which would claw back a factor of 3 at least). I imagine that this could be set _much_ lower in practice (perhaps only going half a sigma more than necessary).
- The disadvantage was the creation of artefacts, since a triangulation along a banana doesn't handle the non-convex negative space very well.

What we should do is to combine the two approaches, namely to have a meshgrid superimposed upon samples for plotting purposes. The only thing you need to be careful with is the `iso_probability_contours_from_samples`, which historically we were able to use by setting the meshgrid weights which are passed in to zero. Previously we used random samples rather than a meshgrid, but an advantage of the meshgrid would be to remove jagged edges in the event the quantile reached the edge.

However, I think we can go one better than random/meshgrid. We definitely want to avoid jagged edges, so having four corner points on boundaries of the plot is a good idea. The meshgrid/random samples could also be in the decorrelated space which scaled_triangulation uses.

The following strategy looks to be a good way of populating the space with plotting points with fewer than 200 samples.

```python
from matplotlib.tri import Triangulation, UniformTriRefiner
from anesthetic import NestedSamples
import numpy as np
import matplotlib.pyplot as plt

samples = NestedSamples(root='./tests/example_data/pc')

n = 100
x = samples.compress(n).x0
y = samples.compress(n).x1
xmin, xmax = samples.x0.quantile([0.00001, 0.99999])
ymin, ymax = samples.x1.quantile([0.00001, 0.99999])

x = (x-xmin)/(xmax-xmin)
y = (y-ymin)/(ymax-ymin)

x = np.concatenate([x, [0,1,0,1]])
y = np.concatenate([y, [0,1,1,0]])
x = np.concatenate([x, np.random.rand(50)])
y = np.concatenate([y, np.random.rand(50)])

tri = Triangulation(x, y)
x = xmin + (xmax-xmin)*x
y = ymin + (ymax-ymin)*y
tri = Triangulation(x, y, tri.triangles)

ax = samples.plot.scatter('x0','x1')
ax.triplot(tri, color='k', lw=0.5)
ax.figure.tight_layout()
```
![tri](https://user-images.githubusercontent.com/3708486/184884390-6096e385-7c50-4d57-9b24-9de4b9def0ff.png)

I think that in principle one could be even smarter than this, e.g. using temperature to spread the plotting samples, but for now this should be better than what we're currently doing.